File: solutions_2453486_0/Python/knzm/tictac.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

def read(f):
    n = int(f.readline().strip())
    for i in xrange(n):
        board = [list(f.readline().strip()) for j in xrange(4)]
        yield board
        # skip empty line
        f.readline()

def get_lines(board):
    for row in board:
        yield row
    for col in zip(*board):
        yield col
    yield [board[i][j] for i, j in [(0, 0), (1, 1), (2, 2), (3, 3)]]
    yield [board[i][j] for i, j in [(0, 3), (1, 2), (2, 1), (3, 0)]]

def solve(board):
    for line in get_lines(board):
        tmp = "".join([c for c in line if c != "T"])
        if all([c == "X" for c in tmp]):
            return "X won"
        elif all([c == "O" for c in tmp]):
            return "O won"
    for row in board:
        if "." in row:
            return "Game has not completed"
    return "Draw"

def main(f):
    for i, board in enumerate(read(f)):
        result = solve(board)
        print "Case #%d: %s" % (i+1, result)

_input = """
6
XXXT
....
OO..
....

XOXT
XXOO
OXOX
XXOO

XOX.
OX..
....
....

OOXX
OXXX
OX.T
O..O

XXXO
..O.
.O..
T...

OXXX
XO..
..O.
...O

""".strip()

_output = """
Case #1: X won
Case #2: Draw
Case #3: Game has not completed
Case #4: O won
Case #5: O won
Case #6: O won
""".strip()

def test_main(compare=False):
    import sys
    from difflib import unified_diff
    from StringIO import StringIO

    if compare:
        stdout = sys.stdout
        sys.stdout = StringIO()
        try:
            main(StringIO(_input))
            result = sys.stdout.getvalue().strip()
        finally:
            sys.stdout = stdout

        print result

        for line in unified_diff(result.splitlines(), _output.splitlines(),
                                 'Output', 'Expect', lineterm=''):
            print line

        if result == _output:
            print "OK"
        else:
            print "NG"

    else:
        main(StringIO(_input))

if __name__ == '__main__':
    test = False
    compare = True
    if test:
        test_main(compare)
    else:
        import sys
        if len(sys.argv) > 1:
            f = open(sys.argv[1])
            main(f)
            f.close()
        else:
            main(sys.stdin)

File: solutions_2453486_1/java/Lacha/ticTac.java
import java.util.*;
	
public class ticTac {
	public static void main(String[] args)
	{
		Scanner in = new Scanner(System.in);
		int numCase = in.nextInt();
		for (int testCase = 0; testCase < numCase; testCase++)
		{
			int[][] board = new int[4][4];	// i is row, j is column, 2 is 'X', -2 is 'O', 100 is 'T', 0 is '.'
			for (int i = 0; i < 4; i++)
			{
				String readLine = in.next();
				for (int j = 0; j < 4; j++)
				{
					if (readLine.charAt(j) == 'X')
						board[i][j] = 2;
					else if (readLine.charAt(j) == 'O')
						board[i][j] = -2;
					else if (readLine.charAt(j) == 'T')
						board[i][j] = 100;
					else board[i][j] = 0;
				}
			}
			int[] row = new int[4];
			int[] column = new int[4];
			int[] diagonal = new int[2];
			boolean filled = true; 
			//add Row, column and diagonal
			boolean gameOver = false;
			for (int i = 0; i < 4; i++)
			{
				for (int j = 0; j < 4; j++)
				{
					if (board[i][j] == 0)
						filled = false;
					row[i] += board[i][j];
					column[j] += board[i][j];
					if (i == j)
					{
						diagonal[0] += board[i][j];
					}
					if (i+j == 3)
					{
						diagonal[1] += board[i][j];
					}
				}
			}
			//test Row, column
			for (int i = 0; i < 4; i++)
			{
				if ((row[i] == 8 || column[i] == 8 || row[i] == 106 || column[i] == 106) && !gameOver)
				{
					System.out.println("Case #" + (testCase+1)+": X won");
					gameOver = true;
				}
				else if ((row[i] == -8 || column[i] == -8 || row[i] == 94 || column[i] == 94) && !gameOver)
				{
					System.out.println("Case #" + (testCase+1)+": O won");
					gameOver = true;
				}
			}
			for (int i = 0; i < 2; i++)
			{
				if ((diagonal[i] == 8 || diagonal[i] == 106) && !gameOver)
				{
					System.out.println("Case #" + (testCase+1)+": X won");
					gameOver = true;
				}
				else if ((diagonal[i] == -8 || diagonal[i] == 94) && !gameOver)
				{
					System.out.println("Case #" + (testCase+1)+": O won");
					gameOver = true;
				}
			}
			if (filled && !gameOver)
			{
				System.out.println("Case #" + (testCase+1)+": Draw");
			}
			else if (!filled && !gameOver){
				System.out.println("Case #" + (testCase+1)+": Game has not completed");
			}
		}
	}
}
